lpeg/io_modules/lpeg/sub_decoder_util.lua

```diff
@@ -0,0 +1,288 @@
+
+*Arguments*
+- sub_decoders (table) sub_decoders configuration table
+```lua
+sub_decoders = {
+-- sub_decoder_name (string) = (string/array) a sub_decoder_name of "*" is
+-- treated as the default when the name does not exist.
+  -- string: decoder or grammar module name
+  -- array: (string and/or array) list of specific messages to parse
+    -- string: Sample message used to locate the correct grammar
+       -- If no grammar matches the sample message then an error is thrown
+       -- and another grammar or module must be added to the printf_messages
+       -- configuration. If multiple grammars match the message, the first
+       -- grammar with the most specific match is selected.
+       -- Note: a special token of `<<DROP>>` and `<<FAIL>>` are reserved for
+       -- the last entry in the array to handle the no match case; <<DROP>>
+       -- silently discards the message and <<FAIL>> reports an error. If
+       -- neither is specified the default no match behavior is to inject the
+       -- original message produced by the syslog decoder.
+    -- array:
+       -- column 1: (string/array)
+          -- string: Sample message (see above)
+          -- array: printf.build_grammar format specification
+       -- column 2: (table/nil)
+          -- Transformation table with Heka message field name keys and a
+          -- value of the fully qualified transformation function name. The
+          -- function returns no values but can error; it receives two
+          -- arguments: the Heka message table and the field name to act on.
+          -- The function can modify the message in any way.
+
+  nginx  = "decoders.nginx.access", -- decoder module name
+  kernel = "lpeg.linux.kernel",     -- grammar module name, must export an lpeg grammar named 'grammar' or 'syslog_grammar'
+  sshd = {
+    -- openssh_portable auth message, imported in printf_messages
+    {"Accepted publickey for foobar from 10.11.12.13 port 4242 ssh2", {remote_addr = "geoip.heka.add_geoip"}},
+  },
+  foo = {
+    "/tmp/input.tsv:23: invalid line", -- custom log defined in printf_messages
+    {{"Status: %s", "status"}, nil},   -- inline printf spec, no transformation
+  },
+}
+```
+- printf_messages (table/nil) see: https://mozilla-services.github.io/lua_sandbox_extensions/lpeg/modules/lpeg/printf.html
+
+*Return*
+- sub_decoders (table)
+
+### load_sub_decoder
+
+Returns the decode function for a single sub_decoder.
+
+*Arguments*
+- sub_decoder (string/table) sub_decoder configuration entry
+```lua
+sub_decoder = "decoders.nginx.access"
+```
+- printf_messages (table/nil) printf_message table (see above)
+
+*Return*
+- decode (function)
+
+### copy_message
+
+Copies a message for use in decoder/subdecoder
+
+*Arguments*
+* src (table) Heka message table. This is a shallow copy of the individual
+  values in the Fields hash and assumes they will be replaced as opposed to
+  modififed when they are tables. The main use of this function is to populate
```

Comment:
modified

Reply:
I much prefer the single direction flow as the default (it is easier to reason about). You get to operate on the message(s) (each stage could go 1-N) fail , modify, inject and/or pass them down and when inject is called they are captured in the system even if the next phase blows up.

However, the current design does not prevent you from implementing different behavior.   To create the 'root' behavior above, the inject_message function could be replaced in the input plugin so all calls would go through the root's inject_message instance where it could do some finalization (e.g. say you wanted to sign everything).